nucypher/blockchain/eth/sol/source/contracts/PREApplication.sol

```diff
@@ -0,0 +1,652 @@
+    */
+    function finishAuthorizationDecrease(address _stakingProvider) external updateReward(_stakingProvider) {
+        StakingProviderInfo storage info = stakingProviderInfo[_stakingProvider];
+        require(info.deauthorizing > 0, "There is no deauthorizing in process");
+        require(info.endDeauthorization <= block.timestamp, "Authorization decrease has not finished yet");
+
+        uint96 toAmount = tStaking.approveAuthorizationDecrease(_stakingProvider);
+
+        if (info.operatorConfirmed) {
+            authorizedOverall -= info.authorized - toAmount;
+        }
+
+        emit AuthorizationDecreaseApproved(_stakingProvider, info.authorized, toAmount);
+        info.authorized = toAmount;
+        info.deauthorizing = 0;
+        info.endDeauthorization = 0;
+
+        if (info.authorized == 0) {
+            _stakingProviderFromOperator[info.operator] = address(0);
+            info.operator = address(0);
+            info.operatorConfirmed == false;
+        }
+    }
+
+    /**
+    * @notice Read authorization from staking contract and store it. Can be called by anyone
+    * @param _stakingProvider Address of staking provider
+    */
+    function resynchronizeAuthorization(address _stakingProvider) external updateReward(_stakingProvider) {
+        StakingProviderInfo storage info = stakingProviderInfo[_stakingProvider];
+        uint96 newAuthorized = tStaking.authorizedStake(_stakingProvider, address(this));
+        require(info.authorized > newAuthorized, "Nothing to synchronize");
+
+        if (info.operatorConfirmed) {
+            authorizedOverall -= info.authorized - newAuthorized;
+        }
+        emit AuthorizationReSynchronized(_stakingProvider, info.authorized, newAuthorized);
+
+        info.authorized = newAuthorized;
+        if (info.authorized < info.deauthorizing) {
+            info.deauthorizing = info.authorized;
+        }
+
+        if (info.authorized == 0) {
+            _stakingProviderFromOperator[info.operator] = address(0);
+            info.operator = address(0);
+            info.operatorConfirmed == false;
+        }
+    }
+
+    //-------------------------Main-------------------------
+    /**
+    * @notice Returns staking provider for specified operator
+    */
+    function stakingProviderFromOperator(address _operator) public view override returns (address) {
+        return _stakingProviderFromOperator[_operator];
+    }
+
+    /**
+    * @notice Returns operator for specified staking provider
+    */
+    function getOperatorFromStakingProvider(address _stakingProvider) public view returns (address) {
+        return stakingProviderInfo[_stakingProvider].operator;
+    }
+
+    /**
+    * @notice Get all tokens delegated to the staking provider
+    */
+    function authorizedStake(address _stakingProvider) public view override returns (uint96) {
+        return stakingProviderInfo[_stakingProvider].authorized;
+    }
```

Comment:
i think this is another function (along with the next 3 - down to line 553) that could be abstracted to a base Application

Reply:
not fully sure about that. `getActiveStakingProviders` is only PRE thing. `getBeneficiary`, `isAuthorized` and `authorizedStake` are mostly used to not read directly from `TokenStaking`. I imagine that we can use same rule as with events (threshold-network/solidity-contracts#88): if it's only for dashboard and we can get this information from `TokenStaking` then don't add to `IApplication`